### DISCUSSION TITLE
Experience with reusable workflows

### DISCUSSION THREAD

**Discussion by jsoref** [community/community#9083] [Actions]
I have an [action](https://github.com/check-spelling/check-spelling), it's a bit complicated, so I've created a [workflow template](https://github.com/check-spelling/spell-check-this/) that uses it.

There's a new feature called reusable workflows, so I [tried to use it](https://github.com/jsoref/spell-check-this/actions).

Other people have already identified pain points:
* One can't easily reference moving versions. -- Specifically, I'd like to be able to have a workflow where users can pass in `@main` or `@prerelease` and not have to maintain two copies of the workflow just for the difference in versions.
* [Some error reporting is unusable](https://github.community/t/github-actions-invalid-value-workflow-reference-references-to-workflows-must-be-rooted-in-github-workflows/210291/4)

Currently, [permissions](https://docs.github.com/en/actions/learn-github-actions/reusing-workflows#supported-keywords-for-jobs-that-call-a-reusable-workflow) are only specifiable by the caller, which means to make my workflow work, I need to write three separate callable workflows:
* [.github/workflows/spell-check.yml](https://github.com/jsoref/check-spelling/blob/reusable/.github/workflows/spell-check.yml)
* [.github/workflows/comment.yml](https://github.com/jsoref/check-spelling/blob/reusable/.github/workflows/comment.yml)
* [.github/workflows/update.yml](https://github.com/jsoref/check-spelling/blob/reusable/.github/workflows/update.yml)

I had to duplicate all of the [inputs from my action.yml](https://github.com/check-spelling/check-spelling/blob/prerelease/action.yml) in each of the reusable workflows. Including copying over all of the default values.

The reusable workflow could go into the action, but,
* it isn't strictly necessary for the action
* if I have it pin to a commit, then each commit it pins to is useless
* if I want to do a release, i need to make a future looking commit on the wrong branch, and then swap the branch out, this is pretty much a disaster, since the whole design is to commit to people that I won't change a release version.

So, maybe it goes into the template repository, but that doesn't make any sense.
So maybe it gets its own repository. But that means it's even harder for me to manage upgrades to inputs, since now I have to actually copy the inputs across repositories.

I'm sure there are uses for reusable workflows, but so far, it doesn't feel like it works for mine.

I'm not sure that people would be more willing to use a reusable workflow that in turn points to a branch, and thus means it can shift out from under them. At least with actions, they can pin a commit if they are concerned. And if I never provided a reusable workflow with pinned shas, they'd never have the ability to pin to a thing.

Personally, having been bitten by potentially pinned versions, I actually prefer branches over shas, although I've already implemented a solution to address the security ramifications of a pinned version.

Can someone talk me through how they'd envision any of this working for a workflow like mine?

**Comment by jsoref** [comment]
Here's a fun rough edge:
https://github.com/jsoref/spell-check-this/actions/runs/1598089694
I expected the comment job to run.

I need the needs, because the workflow might fail for some other reason, and when that happens, I don't want this job to run.

It ran here, but only after I removed an `&& needs...` item.
https://github.com/jsoref/spell-check-this/commit/2f391079d27123baecf2c855cc6ed9f044e67469

Note: the first part of the needs that remained ensured that the other job ran first.

Also note that the needs I didn't check for in this second run was available for the with.

**Comment by ethomson** [comment]
That's a bit of a head scratcher!  I'll ask one of the engineers to take a look at why the `needs.spelling.outputs.artifact_name` didn't evaluate truthily there, but several people are out until the new year.

**Reply by jsoref** [reply]
Happy new year 🥳
